src/util/metadata/metadata_val_traits.rs

```diff
@@ -0,0 +1,208 @@
+use crate::util::Address;
+use core::sync::atomic::*;
+use num_traits::{FromPrimitive, ToPrimitive};
+use num_traits::{Unsigned, WrappingAdd, WrappingSub};
+
+/// Describes bits and log2 bits for the numbers.
+/// If num_traits has this, we do not need our own implementation: https://github.com/rust-num/num-traits/issues/247
+pub trait Bits {
+    const BITS: u32;
+    const LOG2: u32;
+}
+macro_rules! impl_bits_trait {
+    ($t: ty) => {
+        impl Bits for $t {
+            const BITS: u32 = <$t>::BITS;
+            const LOG2: u32 = Self::BITS.trailing_zeros();
+        }
+    };
+}
+impl_bits_trait!(u8);
+impl_bits_trait!(u16);
+impl_bits_trait!(u32);
+impl_bits_trait!(u64);
+impl_bits_trait!(usize);
+
+/// Describes bitwise operations.
+/// If num_traits has this, we do not need our own implementation: https://github.com/rust-num/num-traits/issues/232
+pub trait BitwiseOps {
```

Comment:
Is it sufficient to require the type to implement traits in `std::ops`, so that we don't have to route our own `bitand`, `bitor`, ... to their standard operators?
```rust
use std::ops::*;
pub trait BitwiseOps: BitAnd + BitOr + BitXor + Not {
}
```

Reply:
I cannot get this work. The problem is that ops traits like `BitAnd` have an `Output` associate type, and we cannot set that to our integer/metadata type`MetadataValue`.